File: eventstore/eventstore.go
package eventstore

import (
    "fmt"
    "os"
    "encoding/json"
    "io/ioutil"
	"../domain"
	"time"
)

// here you can change implementation of EventStore from InMem to File
var ins = InMemEventStore { Events: map[string][]domain.Event {}}

func GetInstance() EventStore {
	return ins
}

type EventStore interface {
    SaveEvent(id string, e *domain.Event, v int)
	LoadHistoryFor(id string) []domain.Event
} 

type EventEnvelope struct {
	Id string `json:"id"`
	Time time.Time `json:"time"`
	Title string `json:"title"`
	Type string `json:"type"`
	Version int `json:"version"`
}

const Dir = "/Users/vbaum/Projects/EventStore/Storage/"

type InMemEventStore struct {
	Events map[string][]domain.Event
}

type FileEventStore struct {}

func (mes InMemEventStore) SaveEvent(id string, e *domain.Event, v int) {
    events := mes.LoadHistoryFor(id);

    fmt.Printf("save version %v\n", v)
    fmt.Printf("events %v\n", len(events))

    if len(events) - 1 != v {
        fmt.Printf("Concurrent exception is happen for %v\n", len(events))
        os.Exit(1)
	}

    e.Version = len(events);
    fmt.Printf("event version set to %v\n", e.Version)
	e.Time = time.Now()
	
	mes.Events[id] = append(events, *e)
	fmt.Printf("saved events %v\n", len(mes.Events[id]))
}

func (mes InMemEventStore) LoadHistoryFor(id string) []domain.Event {
	return mes.Events[id]
}

func (fes FileEventStore) SaveEvent(id string, e *domain.Event, v int) {
    events := fes.LoadHistoryFor(id);

    fmt.Printf("save version %v\n", v)
    fmt.Printf("events %v\n", len(events))

    if len(events) - 1 != v {
        fmt.Printf("Concurrent exception is happen for " + string(len(events)))
        os.Exit(1)
	}

    e.Version = len(events);
    fmt.Printf("event version set to %v\n", e.Version)
	e.Time = time.Now()

    put(id, *e);
}

func (fes FileEventStore) LoadHistoryFor(id string) []domain.Event {
	
	path := Dir + string(id) + "/"
	
	events := []domain.Event {}
	
	files, e := ioutil.ReadDir(path)
	
    if e != nil {
        fmt.Printf("Dir read error: %v\n", e)
        
		return events;
    }
	
    for _, f := range files {
		
		event := readFile(path + f.Name())
		events = append(events, event)
    }
	
  	return events;
}

func put(id string, e domain.Event) {
	
	data := EventEnvelope {}
	
	data.Id = string(e.Id)
    fmt.Printf("marshalling %s\n", data.Id)
	data.Time = e.Time
	data.Title = e.Params.(string)
	data.Type = e.Target
	data.Version = e.Version

	dataString, er := json.Marshal(data)
    fmt.Printf("marshalling %s\n", dataString)
	
	if er != nil {
		panic(er)
	}
	
	if os.MkdirAll(fmt.Sprintf("%v%s", Dir, id), 0764) != nil {
		panic("MkdirAll")
	}
	
	err := ioutil.WriteFile(
		fmt.Sprintf("%v%s/%v.json", Dir, id, data.Time.UnixNano() / int64(time.Millisecond)), 
		dataString, 0664)
		
	if err != nil {
        panic(err)
    }
	
    fmt.Printf("writen to flie\n")
	
}

func readFile(path string) domain.Event {
	
	file, e := ioutil.ReadFile(path)

    if e != nil {
		panic(e)
    }
	
	var data EventEnvelope
    json.Unmarshal(file, &data)

	event := domain.Event {}
	event.Id = data.Id
	event.Target = data.Type
	event.Version = data.Version
	
	if data.Title != "" {
		event.Params = data.Title
	}
	
	return event
}

File: lib/EventStore.ex
defmodule EventStore do
	
	def get_path(id) do
		"/Users/vbaum/Projects/EventStore/Storage/#{id}/"
	end
	
	def load_history_for(id) do
		
		path = get_path(id) 
		
		if (! File.exists?(path)) do
			[]
		else
			files = File.ls!(path) |> Enum.filter(fn x -> x != ".DS_Store" end)
			
			Enum.map(files, fn x -> 
			
				{_, c} = File.read("#{path}#{x}")
				IO.puts "file is read #{x}"
				{_, r} = JSON.decode(c)

				%Event{
					id: r["id"],
					raised: r["time"],
					title: r["title"],
					type: String.to_atom(r["type"]),
					version: r["version"]
				}
				
			end)
		end

	end
	
	def to_ticks({mg, s, mc}) do
	    mg*1000*1000*1000 + s * 1000 + mc
	end
	
	def save(e, id, v) do

		events = load_history_for(id)

		if length(events) - 1 != v do
			raise ArgumentError, message: "Concurrent access exception len:#{length(events)} and v:#{v}"
		end
		
		time = to_ticks(:os.timestamp)
		
		event_to_save = %Event{
			id: id, 
			version: length(events), 
			type: e.type, 
			raised: time, 
			title: e.title, 
			completed: e.completed
		}

		{_, content} = JSON.encode([
			id: id,
			time: time,
			title: e.title,
			type: e.type,
			version: length(events)
			])
		
		IO.puts content
		
		dir_path = get_path(id)
		path = "#{dir_path}#{time}.json"
		if File.write!(path, content) do
			IO.puts "file written to #{path}"
		end
		
		
		event_to_save
		
	end
	
	
end